packages/shared/src/hooks/useAuthForms.ts

```diff
@@ -0,0 +1,72 @@
+import { MutableRefObject, useMemo, useRef, useState } from 'react';
+import {
+  FormValues,
+  SocialProviderAccount,
+} from '../components/auth/RegistrationForm';
+import { CloseModalFunc } from '../components/modals/common';
+import { formToJson } from '../lib/form';
+
+interface UseAuthForms {
+  onDiscardCancelled: CloseModalFunc;
+  onDiscardAttempt: CloseModalFunc;
+  onContainerChange: (el: HTMLDivElement) => void;
+  onSocialProviderChange: (account: SocialProviderAccount) => void;
+  container: HTMLDivElement;
+  isDiscardOpen: boolean;
+  socialAccount?: SocialProviderAccount;
+  formRef?: MutableRefObject<HTMLFormElement>;
+}
+
+interface UseAuthFormsProps {
+  onDiscard?: CloseModalFunc;
+}
+
+const useAuthForms = ({ onDiscard }: UseAuthFormsProps = {}): UseAuthForms => {
+  const [container, setContainer] = useState<HTMLDivElement>();
+  const [isDiscardOpen, setIsDiscardOpen] = useState(false);
+  const [socialAccount, setSocialAccount] = useState<SocialProviderAccount>();
+  const formRef = useRef<HTMLFormElement>();
+
+  const onDiscardAttempt: CloseModalFunc = (e) => {
+    if (!formRef?.current) {
+      return onDiscard(e);
+    }
+
+    if (socialAccount) {
+      return onDiscard(e);
+    }
+
+    const { 'traits.email': emailAd, ...rest } = formToJson<FormValues>(
+      formRef.current,
+      {},
+    );
+    const values = Object.values(rest);
+
+    if (values.some((value) => !!value)) {
+      return setIsDiscardOpen(true);
+    }
+
+    return onDiscard(e);
+  };
+
+  const onDiscardCancelled: CloseModalFunc = (e) => {
+    e.stopPropagation();
+    setIsDiscardOpen(false);
+  };
+
+  return useMemo(
+    () => ({
+      onDiscardAttempt,
+      onDiscardCancelled,
+      onContainerChange: setContainer,
+      onSocialProviderChange: setSocialAccount,
+      container,
+      isDiscardOpen,
+      socialAccount,
+      formRef,
+    }),
+    [formRef?.current, container, isDiscardOpen, socialAccount],
```

Comment:
Do we need a `formRef?.current` as a dependency here? Mutable values like ref.current do not triggers re-renders, so result of `useMemo` will be the same if only `formRef?.current` is updated

Reply:
Yes, we don't expect a re-render here, but it would still look for when the dependency has changed its value, which a re-render from the component where this hook was used, happens.

Reply:
Not sure we are on the same place, please check this [codesandox](https://codesandbox.io/s/still-moon-vmbvu1?file=/src/App.tsx) - refs can be safely excluded from dependencies, because React do not track their values

Reply:
Yeah, I understand, but ref values are out of React flow, so they are updated not at the time of rendering (and time of execution of hooks), so you can just remove `ref.current` from dependencies

Some replies from Dan Abramov:
https://github.com/facebook/react/issues/15164
https://twitter.com/dan_abramov/status/1105918794239803392?s=21&t=xdo_Ith7_fYnXDHWYQtnjw

@rebelchris looks like react-hooks ESLint rules do not work for us because react-hooks probably not included in plugins in our config (or maybe it was intentional)

Reply:
I'm not sure you get the point. That's why he said, "In general cases". As I've been saying the whole time, we are not looking for a re-render here explicitly when the refs change. It is a safety net for when the outside state triggered a re-render, it would at least let the hook realize the reference in the dependency could have changed.

Reply:
The point is clear in "In general" cases you don't require dependencies that don't trigger a re-render. That's why it was mentioned it is not part of React flow as it does not trigger re-render when its reference/value changes. And we are not waiting explicitly when that changed. In the end, the dependencies (including the refs) would still be taken a look at if it has changed, which will then re-compute. That's why you'll notice in most of our cases, the refs are not really included, but in this case, it was.